src/auth/digest/Config.cc

```diff
@@ -807,8 +808,15 @@ Auth::Digest::Config::decode(char const *proxy_auth, const char *aRequestRealm)
         switch (t) {
         case DIGEST_USERNAME:
             safe_free(username);
-            if (value.size() != 0)
-                username = xstrndup(value.rawBuf(), value.size() + 1);
+            if (value.size() != 0) {
+                const auto v = value.termedBuf();
+                if (utf8 && !isValidUtf8String(v, v + value.size())) {
+                    auto str = isCP1251EncodingAllowed(request) ? Cp1251ToUtf8(v) : Latin1ToUtf8(v);
+                    username = xstrndup(str.rawContent(), str.length() + 1);
```

Comment:
This xstrndup() call is buggy: xstrndup() has a terrible API that leads to both buggy-looking code like `xstrndup(n_characters, n+1)` and actual bugs like feeding non-terminated rawContent() output to xstrndup() which calls termination-relying length() on its input!

```suggestion
                    username = xstrdup(str.c_str());
```

1. The easiest fix is to use `xstrdup(str.c_str())` as depicted by the suggestion above.
2. An arguably better solution would be to add and use an SBuf::c_strdup() method that has no API and implementation flaws of xstrndup().
3. The arguably best solution would be to fix xstrndup() itself, but that is a lot of work because it requires fixing its API and, hence, reviewing (and possibly adjusting) all 30+ existing callers.

I recommend using either solution 1 or solution 2. Your call. FWIW, I can (slowly) help with solution 2.

P.S. Yes, there is a similar ugly old xstrndup() call that uses rawBuf(). That call only looks buggy because String::rawBuf() output happens to be zero-terminated. And yes, the String::rawBuf() documentation kind of implies otherwise.